test-suite/src/test/scala/org/scalajs/testsuite/compiler/OptimizerTest.scala

```diff
@@ -65,6 +65,20 @@ object OptimizerTest extends JasmineTest {
       expect(elements).toEqual(js.Array(10, 9, 8, 7, 6, 5, 4, 3, 2))
     }
 
+    it("must not break classOf[T] == classOf[U] - #1658") {
+      expect(classOf[String] == classOf[String]).toBeTruthy
+      expect(classOf[Int] == classOf[Int]).toBeTruthy
+      expect(classOf[Array[Int]] == classOf[Array[Int]]).toBeTruthy
+      expect(classOf[Array[String]] == classOf[Array[String]]).toBeTruthy
+
+      expect(classOf[String] == classOf[Int]).toBeFalsy
+      expect(classOf[Seq[_]] == classOf[List[_]]).toBeFalsy
```

Comment:
What about `classOf[Long] == classOf[Int]` and other classes that end up using the js primitive representation.